File: 구름 알고리즘 미션/그래프 탐색 응용/맛집 탐방/맛집 탐방.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const N = parseInt(input[0]);
const graph = Array.from({ length: N + 1 }, () => []);

for (let i = 1; i < N; i++) {
  const [u, v] = input[i].split(" ").map(Number);
  graph[u].push(v);
  graph[v].push(u);
}

function dfs(start) {
  const visited = Array(N + 1).fill(false);
  const stack = [[start, 0]];
  let farNode = start;
  let maxDist = 0;

  while (stack.length > 0) {
    const [node, dist] = stack.pop();

    if (visited[node]) continue;
    visited[node] = true;

    if (dist > maxDist) {
      maxDist = dist;
      farNode = node;
    }

    for (const next of graph[node]) {
      if (!visited[next]) {
        stack.push([next, dist + 1]);
      }
    }
  }

  return [farNode, maxDist];
}

const [nodeA] = dfs(1);
const [, diameter] = dfs(nodeA);

console.log(diameter + 1);

File: 구름 알고리즘 미션/그래프 탐색 응용/맛집 탐방/맛집 탐방.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N = int(input())
graph = [[] for _ in range(N + 1)]

for _ in range(N - 1):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

def dfs(node, dist, visited):
    visited[node] = True
    far_node = node
    max_dist = dist

    for next_node in graph[node]:
        if not visited[next_node]:
            temp_node, temp_dist = dfs(next_node, dist + 1, visited)
            if temp_dist > max_dist:
                far_node, max_dist = temp_node, temp_dist

    return far_node, max_dist

visited = [False] * (N + 1)
far_node, _ = dfs(1, 0, visited)

visited = [False] * (N + 1)
_, max_length = dfs(far_node, 0, visited)

print(max_length + 1)
